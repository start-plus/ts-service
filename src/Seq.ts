export class Seq {
  private id = 0;

  getNext() {
    return ++this.id;
  }

  reset() {
    this.id = 0;
  }
}
