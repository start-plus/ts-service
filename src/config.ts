import * as Logger from 'bunyan';

export interface Config {
  removeFields: string[];
  debug: boolean;
  depth: number;
  maxArrayLength: number;
  loggerFactory: (serviceName: string) => Logger;
}

export function getDefaultConfig(): Config {
  return {
    removeFields: ['password', 'token', 'accessToken'],
    debug: true,
    depth: 4,
    maxArrayLength: 30,
    loggerFactory(serviceName) {
      return Logger.createLogger({
        name: serviceName,
        level: this.debug ? 'debug' : 'error',
      });
    },
  };
}

export const globalConfig = getDefaultConfig();

export function configure(config: Partial<Config>) {
  Object.assign(globalConfig, config);
}
