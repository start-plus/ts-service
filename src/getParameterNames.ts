/*
 From
 https://github.com/goatslacker/get-parameter-names
 and 
 https://www.npmjs.com/package/@avejidah/get-parameter-names
 */

var COMMENTS = /((\/\/.*$)|(\/\*[\s\S]*?\*\/))/gm;
var DEFAULT_PARAMS = /=[^,]+/gm;
var FAT_ARROWS = /=>.*$/gm;
var SPACES = /\s/gm;
var BEFORE_OPENING_PAREN = /^[^(]*\(/gm;
var AFTER_CLOSING_PAREN = /^([^)]*)\).*$/gm;

function joinDestructors(args: string[]) {
  let count = 0;
  const ret: string[] = [];
  let current: string[] = [];
  args.forEach(param => {
    if (param.includes('{')) {
      count++;
      current.push(param);
    } else if (param.includes('}')) {
      count--;
      current.push(param);
      if (!count) {
        ret.push(current.join(', '));
        current = [];
      }
    } else {
      ret.push(param);
    }
  });
  return ret;
}

export function getParameterNames(fn: Function) {
  var code = fn
    .toString()
    .replace(SPACES, '')
    .replace(COMMENTS, '')
    .replace(FAT_ARROWS, '')
    .replace(DEFAULT_PARAMS, '')
    .replace(BEFORE_OPENING_PAREN, '')
    .replace(AFTER_CLOSING_PAREN, '$1');

  return joinDestructors(code ? code.split(',') : []);
}
