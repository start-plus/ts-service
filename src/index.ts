export * from './decorators';
export * from './config';
