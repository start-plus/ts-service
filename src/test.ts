import * as Joi from 'joi';
import {} from 'class-validator';
import 'reflect-metadata';

const fieldMetadataKey = Symbol('field');
const schemaMetadataKey = Symbol('schema');

function Field(a: any) {
  console.log('Field(): evaluated');
  return function(
    target: any,
    propertyKey: string,
    // descriptor: PropertyDescriptor,
  ) {
    console.log('Field(): called');
    // console.log({ target, propertyKey, FooOptions });
    // Reflect.defineMetadata(fieldMetadataKey, a, target.constructor, propertyKey);
    const meta =
      Reflect.getMetadata(
        schemaMetadataKey,
        target.constructor,
        schemaMetadataKey,
      ) || {};
    meta[propertyKey] = a;
    Reflect.defineMetadata(
      fieldMetadataKey,
      a,
      target.constructor,
      schemaMetadataKey,
    );
    // console.log('meta field', meta);
  };
}

function Method(a?: any) {
  console.log('Field(): evaluated');
  return function(
    target: any,
    propertyKey: string,
    descriptor: PropertyDescriptor,
  ) {
    let paramtypes = Reflect.getMetadata(
      'design:paramtypes',
      target,
      propertyKey,
    );
    console.log('Method', { target, propertyKey });
    console.log(paramtypes);
    // const meta = Reflect.getMetadata(fieldMetadataKey, FooOptions, 'name');
    // console.log('type.name', type.name);
    // console.log('meta', meta);
    // console.log('Field(): called');
    // console.log(new FooOptions() instanceof type);
    // console.log(typeof type);
    // console.log(type.toString());
    // console.log(type().toString());
    // console.log(new (type())());
    // console.log('xxxx');
  };
}
function Options(a?: any) {
  console.log('Options(): evaluated');
  return function(
    target: Function,
    // propertyKey: string,
    // descriptor: PropertyDescriptor,
  ) {};
}

function Service(a?: any) {
  console.log('Service(): evaluated');
  return function(
    target: Function,
    // propertyKey: string,
    // descriptor: PropertyDescriptor,
  ) {
    console.log('Service(): called', target.name);
    for (const name of Object.getOwnPropertyNames(target.prototype)) {
      if (name === 'constructor') {
        continue;
      }
      const method = target.prototype[name];
      console.log(name);

      console.log('Service', { target: target.prototype, propertyKey: name });
      let paramtypes = Reflect.getOwnMetadata(
        'design:paramtypes',
        target.prototype,
        name,
      );
      console.log(paramtypes);
      // console.log(name, method);
      // console.log(Object.getOwnPropertyNames(method));
    }
    // console.log(target.prototype.foo);
  };
}

// function Param() {
//   return function(
//     target: Object,
//     propertyKey: string | symbol,
//     parameterIndex: number,
//   ) {
//     console.log('Param(): called');
//     let type = Reflect.getOwnMetadata('design:type', target, propertyKey);
//     console.log(Reflect.getMetadataKeys(target));
//     const str: any = new FooOptions();
//     console.log(str instanceof type);
//     // console.log(type('as'));
//     // for (const name of Object.getOwnPropertyNames(target.prototype)) {
//     //   if (name === 'constructor') {
//     //     continue;
//     //   }
//     //   const method = target.prototype[name];
//     //   console.log(name, method);
//     //   console.log(Object.getOwnPropertyNames(method));
//     // }
//     // console.log(target.prototype.foo);
//   };
// }

@Options()
class FooOptions {
  private constructor() {}

  @Field(Joi.string().required())
  name: string;

  @Field(Joi.number().required())
  id: number;

  @Field(Joi.object().required())
  nested: {
    address: string;
  };

  @Field(Joi.object().required())
  bar?: object;

  @Field(Joi.object().required())
  bar2: object | null;

  @Field(Joi.object().required())
  bar3: number | string;
}

const test: FooOptions = {
  name: 'foo',
  id: 123,
  nested: { address: 'aa' },
  bar2: null,
  bar3: 12,
};

class OtherOptions<T> {
  @Field(
    Joi.array()
      .items(Joi.string())
      .required(),
  )
  items: T[];
}

@Service()
class MyService {
  @Method()
  async foo(options: FooOptions) {}

  // @Method()
  // otherMethod(a: OtherOptions<string>, b: string) {
  //   return 123;
  // }
}

const f = Joi.object().required();
if (f instanceof Joi.Obj) {
}

// class Point {
//   x: number;
//   y: number;
// }

// class Line {
//   private _p0: Point;
//   private _p1: Point;

//   @validate
//   set p0(value: Point) {
//     this._p0 = value;
//   }
//   get p0() {
//     return this._p0;
//   }

//   @validate
//   set p1(value: Point) {
//     this._p1 = value;
//   }
//   get p1() {
//     return this._p1;
//   }
// }

// function validate<T>(
//   target: any,
//   propertyKey: string,
//   descriptor: TypedPropertyDescriptor<T>,
// ) {
//   let set = descriptor.set;
//   descriptor.set = function(value: T) {
//     let type = Reflect.getMetadata('design:type', target, propertyKey);
//     console.log('validate type', type);
//     if (!(value instanceof type)) {
//       throw new TypeError('Invalid type.');
//     }
//     set(value);
//   };
// }

// const l = new Line();
// l.p0 = 'abc' as any;

const def = {
  foo: {
    type: 'string',
    joi: Joi.string(),
  },
};

type Def = {
  [s: string]: {
    type: 'string' | 1;
    joi: Joi.StringSchema;
  };
};

function getType(obj: Def) {}

function guess(a: number) {
  if (a === 1) {
    return 'foo' as string;
  }
  return 100 as number;
}

const ret = guess(1);
