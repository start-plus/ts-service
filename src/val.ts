// import {
//   validate,
//   Contains,
//   IsInt,
//   Length,
//   IsEmail,
//   IsFQDN,
//   IsDate,
//   Min,
//   Max,
//   ValidateNested,
// } from 'class-validator';
// import { transformAndValidateSync } from 'class-transformer-validator';

// export class Author {
//   @Min(0)
//   id: number;

//   name: string;
// }

// export class Post {
//   @Length(10, 20)
//   title: string;

//   // @Contains('hello') text: string;

//   // @IsInt()
//   // @Min(0)
//   // @Max(10)
//   // rating: number;

//   // @IsEmail() email: string;

//   // @IsFQDN() site: string;

//   // @IsDate() createDate: Date;

//   @ValidateNested() author: Author;
// }

// let post = new Post();
// post.title = 'Hello'; // should not pass
// // post.text = 'this is a great post about hell world'; // should not pass
// // post.rating = 11; // should not pass
// // post.email = 'google.com'; // should not pass
// // post.site = 'googlecom'; // should not pass
// post.author = { name: 123 } as Author;

// try {
//   transformAndValidateSync(Post, { title: 'a', author: { name: 123 } });
// } catch (e) {
//   console.log(e);
// }
// // validate(post, { validationError: { target: false } }).then(errors => {
// //   // errors is an array of validation errors
// //   if (errors.length > 0) {
// //     console.log('validation failed. errors: ', errors);
// //   } else {
// //     console.log('validation succeed');
// //   }
// // });
