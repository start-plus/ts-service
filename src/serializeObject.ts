import * as util from 'util';
import { Config } from './config';

/**
 * Remove invalid properties from the object and hide long arrays
 * @param  obj the object
 * @returns the new object with removed properties
 * @private
 */
function _sanitizeObject(config: Config, obj: any) {
  if (obj === undefined) {
    return obj;
  }
  const seen: any[] = [];
  return JSON.parse(
    JSON.stringify(obj, (name, value) => {
      if (seen.indexOf(value) !== -1) {
        return '[Circular]';
      }
      if (value != null && typeof value === 'object') {
        seen.push(value);
      }
      // Array of field names that should not be logged
      // add field if necessary (password, tokens etc)
      if (config.removeFields.indexOf(name) !== -1) {
        return '<removed>';
      }
      if (name === 'req' && value && value.connection) {
        return {
          method: value.method,
          url: value.url,
          headers: value.headers,
          remoteAddress: value.connection.remoteAddress,
          remotePort: value.connection.remotePort,
        };
      }
      if (name === 'res' && value && value.statusCode) {
        return {
          statusCode: value.statusCode,
          header: value._header,
        };
      }
      if (Array.isArray(value) && value.length > config.maxArrayLength) {
        return `Array(${value.length})`;
      }
      return value;
    }),
  );
}

export function serializeObject(config: Config, obj: any) {
  return util.inspect(_sanitizeObject(config, obj), { depth: config.depth });
}
