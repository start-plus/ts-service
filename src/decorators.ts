import * as Joi from 'joi';
import { getParameterNames } from './getParameterNames';

const serviceMetadataKey = Symbol('service');
const schemaMetadataKey = Symbol('schema');
const validateKey = Symbol('validate');

export function schema(joiSchema: Joi.Schema) {
  return function(target: any, propertyKey?: string, parameterIndex?: number) {
    const metadataKey = propertyKey ? parameterIndex : schemaMetadataKey;
    Reflect.defineMetadata(metadataKey, joiSchema, target, schemaMetadataKey);
  };
}

export function service(target: any) {
  for (const name of Object.getOwnPropertyNames(target.prototype)) {
    if (name === 'constructor') {
      continue;
    }
    const method = target.prototype[name];
    const validate = Reflect.getMetadata(validateKey, target.prototype, name);
    if (validate == null) {
      throw new Error(
        `Method "${
          target.name
        }.${name}" should be annotated with @validate or @ignore`,
      );
    }
    if (!validate) {
      return;
    }
  }
}

export function validate(
  target: any,
  propertyKey: string,
  descriptor: PropertyDescriptor,
) {
  Reflect.defineMetadata(validateKey, true, target, propertyKey);
  const params =
    Reflect.getMetadata('design:paramtypes', target, propertyKey) || [];
  const method = target[propertyKey];
  const paramNames = getParameterNames(method);
  params.forEach((type: Function, i: number) => {
    const schema =
      // inline schema
      Reflect.getMetadata(i, target, schemaMetadataKey) ||
      // type schema
      Reflect.getMetadata(schemaMetadataKey, type, schemaMetadataKey);
    if (!schema) {
      throw new Error(
        `Parameter ${paramNames[i]} in ${
          target.constructor.name
        }.${propertyKey} is missing @schema annotation`,
      );
    }
  });
}

export function ignore(
  target: any,
  propertyKey: string,
  descriptor: PropertyDescriptor,
) {
  Reflect.defineMetadata(validateKey, false, target, propertyKey);
}

// @schema(Joi.object())
// class Example {
//   prop1: string;
// }

// @service
// class MyService {
//   @validate
//   searchUsers(criteria: Example) {}

//   @validate
//   getUserById(
//     @schema(Joi.string().required())
//     id: string,
//   ) {}
// }
