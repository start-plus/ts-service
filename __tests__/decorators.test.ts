import 'reflect-metadata';
import * as Joi from 'joi';
import { service, validate, ignore, schema } from '../src/decorators';

describe('decorators', () => {
  describe('validation', () => {
    it('should throw if method is missing @validate', () => {
      expect(() => {
        @service
        class MyService {
          foo() {
            return 10;
          }
        }
      }).toThrowErrorMatchingSnapshot();
    });

    it('should not throw if method is annotated with @validate', () => {
      expect(() => {
        @service
        class MyService {
          @validate
          foo() {
            return 10;
          }
        }
      }).not.toThrow();
    });

    it('should not throw if method is annotated with @ignore', () => {
      expect(() => {
        @service
        class MyService {
          @ignore
          foo() {
            return 10;
          }
        }
      }).not.toThrow();
    });

    it('should throw if param is missing @schema annotation', () => {
      expect(() => {
        class InvalidOptions {
          id: number;
        }

        @service
        class MyService {
          @validate
          foo(options: InvalidOptions) {
            return 10;
          }
        }
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw if 2nd param is missing @schema annotation', () => {
      expect(() => {
        @schema(Joi.object())
        class ValidOptions {
          id: number;
        }
        class InvalidOptions {
          id: number;
        }

        @service
        class MyService {
          @validate
          foo(options1: ValidOptions, options2: InvalidOptions) {
            return 10;
          }
        }
      }).toThrowErrorMatchingSnapshot();
    });

    it('should not throw with annotated with inline @schema', () => {
      expect(() => {
        @service
        class MyService {
          @validate
          foo(
            @schema(Joi.number().required())
            id: number,
          ) {
            return id + 10;
          }
        }
      }).not.toThrow();
    });
  });

  describe('functional', () => {
    it('validate sync function', () => {});
  });
});
