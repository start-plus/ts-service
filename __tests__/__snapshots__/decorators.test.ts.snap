// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators validation should throw if 2nd param is missing @schema annotation 1`] = `"Parameter options2 in MyService.foo is missing @schema annotation"`;

exports[`decorators validation should throw if method is missing @validate 1`] = `"Method \\"MyService.foo\\" should be annotated with @validate or @ignore"`;

exports[`decorators validation should throw if param is missing @schema annotation 1`] = `"Parameter options in MyService.foo is missing @schema annotation"`;
